lib/types.nix

```diff
@@ -107,22 +107,80 @@ rec {
       merge = mergeEqualOption;
     };
 
+### INTEGERS
+
     int = mkOptionType rec {
-      name = "int";
-      description = "integer";
-      check = isInt;
-      merge = mergeOneOption;
-    };
+        name = "int";
+        description = "signed integer";
+        check = isInt;
+        merge = mergeOneOption;
+      };
+
+    # specialized subdomains of int
+    ints =
+      let
+        betweenDesc = lowest: highest:
+          "${toString lowest} and ${toString highest} (both inclusive)";
+        between = lowest: highest: assert lowest <= highest;
+          addCheck int (x: x >= lowest && x <= highest) // {
+            name = "intBetween";
+            description = "integer between ${betweenDesc lowest highest}";
+          };
+        ign = lowest: highest: name: docStart:
+          between lowest highest // {
+            inherit name;
+            description = docStart + "; between ${betweenDesc lowest highest}";
+          };
+        unsign = bit: range: ign 0 (range - 1)
+          "unsignedInt${toString bit}" "${toString bit} bit unsigned integer";
+        sign = bit: range: ign (0 - (range / 2)) (range / 2 - 1)
+          "signedInt${toString bit}" "${toString bit} bit signed integer";
+
+      in rec {
+        /* An int with a fixed range.
+
+          Example:
+            (ints.between 0 100).check (-1)
+            => false
+            (ints.between 0 100).check (101)
+            => false
+            (ints.between 0 0).check 0
+            => true
+        */
+        inherit between;
+
+        unsigned = addCheck types.int (x: x >= 0) // {
+          name = "unsignedInt";
+          description = "unsigned integer, meaning >=0";
+        };
+        positive = addCheck types.int (x: x > 0) // {
+          name = "positiveInt";
+          description = "positive integer, meaning >0";
+        };
+        u8 = unsign 8 256;
+        u16 = unsign 16 65536;
+        u32 = unsign 32 4294967296;
+        # the biggest int the nix lexer accepts is 9223372036854775808
+        # the smallest int the nix lexer accepts is -9223372036854775807
+        # unsigned64 = unsign 64 18446744073709551616;
+
+        s8 = sign 8 256;
+        s16 = sign 16 65536;
+        s32 = sign 32 4294967296;
+      };
 
+### STRINGS
+
+    /* Basic string type. Use this, not `string'. */
     str = mkOptionType {
       name = "str";
       description = "string";
       check = isString;
       merge = mergeOneOption;
     };
 
-    # Merge multiple definitions by concatenating them (with the given
-    # separator between the values).
+    /* Merge multiple definitions by concatenating them (with the given
+       separator between the values). */
```

Comment:
nit: Avoid /* */ comments for small descriptions.

Reply:
I’d like to make `/**/` default for “docstrings”, so that they can be distinguished from in-line comments easily.